man/man3/ftw.3
.\" @(#)ftw.3 1.1 92/07/30 SMI; from S5R2
.TH FTW 3 "22 November 1987"
.SH NAME
ftw \- walk a file tree
.SH SYNOPSIS
.LP
.nf
.ft B
#include <ftw.h>
.ft
.fi
.LP
.nf
.ft B
int ftw(path, fn, depth)
char *path;
int (*fn)(\|);
int depth;
.ft
.fi
.SH DESCRIPTION
.IX ftw() "" "\fLftw()\fR \(em traverse file tree"
.IX file "\fLftw()\fR \(em traverse file tree"
.LP
.B ftw(\|)
recursively descends the directory hierarchy rooted in
.IR path .
For each object in the hierarchy,
.B ftw(\|)
calls
.IR fn ,
passing it a pointer to a
null-terminated
character string containing the name of the object, a pointer to a
.B stat(\|)
structure (see
.BR stat (2V))
containing information about the object, and an integer.
Possible values of the integer, defined in the
.B <ftw.h>
header file, are
.SM FTW_F\s0
for a file,
.SM FTW_D\s0
for a directory,
.SM FTW_DNR\s0
for a directory that cannot be read, and
.SM FTW_NS\s0
for an object for which
.B stat(\|)
could not successfully be executed.
If the integer is
.SM FTW_DNR\s0,
descendants of that directory will not be processed.
If the integer is
.SM FTW_NS\s0,
the
.B stat(\|)
structure will contain garbage.
An example of an object that would cause
.SM FTW_NS\s0
to be passed to
.I fn
would be a file in a directory
with read but without execute (search) permission.
.LP
.B ftw(\|)
visits a directory before visiting any of its descendants.
.LP
The tree traversal continues until the tree is exhausted,
an invocation of
.I fn
returns a nonzero value,
or some error is detected within
.B ftw(\|)
(such as an I/O error).
If the tree is exhausted,
.B ftw(\|)
returns zero.  If
.I fn
returns a nonzero value,
.B ftw(\|)
stops its tree traversal and returns whatever
value was returned by
.IR fn .
If
.B ftw(\|)
detects an error, it returns
\-1, and sets the error type in
.BR errno .
.LP
.B ftw(\|)
uses one file descriptor for each level in the tree.
The
.I depth
argument limits the number of file descriptors so used.
If
.I depth
is zero or negative, the effect is the same as if it were 1.
.I depth
must not be greater than the number of file descriptors currently
available for use.
.B ftw(\|)
will run more quickly if
.I depth
is at least as large as the number of levels in the tree.
.SH SEE ALSO
.BR stat (2V),
.BR malloc (3V)
.SH BUGS
Because
.B ftw(\|)
is recursive, it is possible for it to terminate with a memory
fault when applied to very deep file structures.
.LP
It could be made to run faster and use less storage on deep
structures at the cost of considerable complexity.
.LP
.B ftw(\|)
uses
.BR malloc (3V)
to allocate dynamic storage during its operation.
If
.B ftw(\|)
is forcibly terminated, such as by
.B longjmp(\|)
being executed by
.I fn
or an interrupt routine,
.B ftw(\|)
will not have a chance to free that storage,
so it will remain permanently allocated.
A safe way to handle interrupts is to store
the fact that an interrupt has occurred,
and arrange to have
.I fn
return a nonzero value at its next invocation.


lib/libc/gen/common/ftw.c
#if !defined(lint) && defined(SCCSIDS)
static	char sccsid[] = "@(#)ftw.c 1.1 92/07/30 SMI"; /* from S5R2 1.2 */
#endif

/*LINTLIBRARY*/
/***************************************************************
 *	ftw - file tree walk
 *
 *	int ftw (path, fn, depth)  char *path; int (*fn)(); int depth;
 *
 *	Given a path name, ftw starts from the file given by that path
 *	name and visits each file and directory in the tree beneath
 *	that file.  If a single file has multiple links within the
 *	structure, it will be visited once for each such link.
 *	For each object visited, fn is called with three arguments.
 *	The first contains the path name of the object, the second
 *	contains a pointer to a stat buffer which will usually hold
 *	appropriate information for the object and the third will
 *	contain an integer value giving additional information about
 *
 *		FTW_F	The object is a file for which stat was
 *			successful.  It does not guarantee that the
 *			file can actually be read.
 *
 *		FTW_D	The object is a directory for which stat and
 *			open for read were both successful.
 *
 *		FTW_DNR	The object is a directory for which stat
 *			succeeded, but which cannot be read.  Because
 *			the directory cannot be read, fn will not be
 *			called for any descendants of this directory.
 *
 *		FTW_NS	Stat failed on the object because of lack of
 *			appropriate permission, or because the object is a 
 *			symbolic link that points to a non-existent file.
 *			This indication will be given, for example, for each 
 *			file in a directory with read but no execute 
 *			permission.   Because stat failed, it is not 
 *			possible to determine whether this object is a file 
 *			or a directory.  The stat buffer passed to fn will
 *			contain garbage.  Stat failure for any reason
 *			other than lack of permission will be
 *			considered an error and will cause ftw to stop 
 *			and return -1 to its caller.
 *
 *	If fn returns nonzero, ftw stops and returns the same value
 *	to its caller.  If ftw gets into other trouble along the way,
 *	it returns -1 and leaves an indication of the cause in errno.
 *
 *	The third argument to ftw does not limit the depth to which
 *	ftw will go.  Rather, it limits the depth to which ftw will
 *	go before it starts recycling file descriptors.  In general,
 *	it is necessary to use a file descriptor for each level of the
 *	tree, but they can be recycled for deep trees by saving the
 *	position, closing, re-opening, and seeking.  It is possible
 *	to start recycling file descriptors by sensing when we have
 *	run out, but in general this will not be terribly useful if
 *	fn expects to be able to open files.  We could also figure out
 *	how many file descriptors are available and guarantee a certain
 *	number to fn, but we would not know how many to guarantee,
 *	and we do not want to impose the extra overhead on a caller who
 *	knows how many are available without having to figure it out.
 *
 *	It is possible for ftw to die with a memory fault in the event
 *	of a file system so deeply nested that the stack overflows.
 **************************************************************/

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/dir.h>
#include <errno.h>
#include <ftw.h>

#define NULL 0

extern char *malloc(), *strcpy();
extern void free();
extern int errno;

int
ftw(path, fn, depth)
char *path;
int (*fn)();
int depth;
{
	int rc, n;
	DIR *dirp;
	char *subpath, *component;
	struct stat sb;
	struct direct *dp;

	/* Try to get file status.
				If unsuccessful, errno will say why. */
	if(stat(path, &sb) < 0) {
		if (errno == EACCES) {
			return((*fn)(path, &sb, FTW_NS));
		} else if (errno == ENOENT) {
			/* Check if symbolic link points to non-existent file */
			if (lstat(path, &sb) < 0) {
				return(-1);
			}
			else if ((sb.st_mode & S_IFMT) == S_IFLNK) {
				errno = ENOENT;
				return((*fn)(path, &sb, FTW_NS));
			}
			else {
				return(-1);
			}
		} else {
			return(-1);
		}
	}

	/*
	 *	The stat succeeded, so we know the object exists.
	 *	If not a directory, call the user function and return.
	 */
	if((sb.st_mode & S_IFMT) != S_IFDIR)
		return((*fn)(path, &sb, FTW_F));

	/*
	 *	The object was a directory.
	 *
	 *	Open a file to read the directory
	 */
	dirp = opendir(path);

	/*
	 *	Call the user function, telling it whether
	 *	the directory can be read.  If it can't be read
	 *	call the user function or indicate an error,
	 *	depending on the reason it couldn't be read.
	 */
	if(dirp == NULL)
		return(errno == EACCES? (*fn)(path, &sb, FTW_DNR): -1);

	/* We could read the directory.  Call user function. */
	rc = (*fn)(path, &sb, FTW_D);
	if(rc != 0)
		return(rc);

	/* Allocate a buffer to hold generated pathnames. */
	n = strlen(path);
	subpath = malloc((unsigned)(n+MAXNAMLEN+2));
	if(subpath == NULL) {
		closedir(dirp);
		errno = ENOMEM;
		return(-1);
	}
	
	/* Create a prefix to which we will append component names */
	(void)strcpy(subpath, path);
	if(subpath[0] != '\0' && subpath[n-1] != '/')
		subpath[n++] = '/';
	component = &subpath[n];

	/*
	 *	Read the directory one component at a time.
	 *	We must ignore "." and "..", but other than that,
	 *	just create a path name and call self to check it out.
	 */
	while((dp = readdir(dirp)) != NULL) {
		if(strcmp(dp->d_name, ".") != 0 &&
			strcmp(dp->d_name, "..") != 0) {
				long here;

			/* Append component name to the working path */
			(void)strcpy(component, dp->d_name);

			/*
			 *	If we are about to exceed our depth,
			 *	remember where we are and close a file.
			 */
			if(depth <= 1) {
				here = telldir(dirp);
				closedir(dirp);
			}

			/*
			 *	Do a recursive call to process the file.
			 *	(watch this, sports fans)
			 */
			rc = ftw(subpath, fn, depth-1);
			if(rc != 0) {
				free(subpath);
				if(depth > 1)
					closedir(dirp);
				return(rc);
			}

			/*
			 *	If we closed the file, try to reopen it.
			 */
			if(depth <= 1) {
				dirp = opendir(path);
				if(dirp == NULL) {
					free(subpath);
					return(-1);
				}
				seekdir(dirp, here);
			}
		}
	}

	/*
	 *	We got out of the subdirectory loop.  The return from
	 *	the final readdir is in dp.  Clean up.
	 */
	free(subpath);
	closedir(dirp);
	return(0);
}
